src/psyclone/psyir/symbols/symbol_table.py

```diff
@@ -1087,11 +1158,64 @@ def swap(self, old_symbol, new_symbol):
             raise SymbolError(
                 f"Cannot swap symbols that have different names, got: "
                 f"'{old_symbol.name}' and '{new_symbol.name}'")
-        # TODO #898 remove() does not currently check for any uses of
-        # old_symbol.
+        self._replace_symbol_refs(old_symbol, new_symbol)
+        if self.node:
+            table = SymbolTable()
+            table.add(new_symbol)
+            self.node.replace_symbols_using(table)
         self.remove(old_symbol)
         self.add(new_symbol)
 
+    def _replace_symbol_refs(self, old_sym: Symbol, new_sym: Symbol):
+        '''
+        Looks through all Symbols referenced in the *definitions* of the
+        Symbols in this table and replaces any instances of `old_sym` with
+        `new_sym`.
+
+        Note, this method does not attempt to update any PSyIR tree that may
+        be associated with this table. That should be done separately using
+        Node.replace_symbols_using().
+
+        :param old_sym: the existing symbol to replace.
+        :param new_sym: the replacement symbol.
+
+        :raises InternalError: if an access is found in an unexpected type
+            of Node.
+
+        '''
+        norm_name = SymbolTable._normalize(old_sym.name)
+        from psyclone.core import Signature
+        from psyclone.core.variables_access_info import VariablesAccessInfo
+        from psyclone.psyir.nodes import CodeBlock, Literal, Reference
+        vai = VariablesAccessInfo()
+        self.reference_accesses(vai)
+        sig = Signature(norm_name)
+        if sig not in vai:
+            return
+        new_table = SymbolTable()
+        new_table.add(new_sym)
+        for access in vai[sig].all_accesses:
+            if isinstance(access.node, Symbol):
+                sym = access.node
+                sym.replace_symbols_using(new_table)
+            elif isinstance(access.node, Reference):
+                access.node.symbol = new_sym
+            elif isinstance(access.node, Literal):
+                oldtype = access.node.datatype
+                newtype = ScalarType(oldtype.intrinsic,
+                                     new_sym)
+                access.node.replace_with(
+                    Literal(access.node.value, newtype))
+            elif isinstance(access.node, CodeBlock):
+                # Nothing to do here as a CodeBlock does not contain
+                # Symbols (just a parse tree).
+                pass
```

Comment:
It does contain symbols: `codeblock.get_symbols_names()`. Should be fail if old_sym.name is found in this list?

Reply:
No...but the reason why not was not obvious: this method expects the two Symbols to have the same name. I've added a check for this, updated the docstring and extended this comment.